### PR TITLE
Updates to improve the performance and logging of the transformer inf…

### PR DESCRIPTION
**Pull request by eelstretching**
…rastructure

### Description

The performance of the transformers was suboptimal when there was a large number of features. Got the transformation time for a large text classification problem down from more than two hours to less than two minutes. Along the way, I added some logging (at Level.FINE) so you can see what's happening, if you choose to do so.

As I was working on a text classification problem, I added and IDFTransformation that can set TFIDF weights in the example.

### Motivation

Transformer performance was far too slow in some cases.

### CODE REVIEW

File: Core/src/main/java/org/tribuo/impl/ArrayExample.java [hunk 1]
```diff
@@ -159,12 +159,19 @@ public ArrayExample(T output, List<? extends Feature> features) {
...
         super(other);
-        featureNames = new String[other.size()];
-        featureValues = new double[other.size()];
-        for (Feature f : other) {
-            featureNames[size] = f.getName();
-            featureValues[size] = f.getValue();
-            size++;
+        if(other instanceof ArrayExample) {
+            ArrayExample< T> otherArr = (ArrayExample<T>) other;
+            featureNames = Arrays.copyOf(otherArr.featureNames, otherArr.featureNames.length);
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/impl/ArrayExample.java:164]: I know I wrote this, but the Arrays.copyOf should probably be to `size` not `otherArr.featureNames.length` to make it match up with the other side of the else. Plus it's slightly more memory efficient that way.

File: Core/src/main/java/org/tribuo/impl/ArrayExample.java [hunk 2]
```diff
@@ -408,14 +415,33 @@ public void set(Feature feature) {
...
     @Override
     public void transform(TransformerMap transformerMap) {
-        for (Map.Entry<String,List<Transformer>> e : transformerMap.entrySet()) {
-            int index = Arrays.binarySearch(featureNames,0,size,e.getKey());
-            if (index >= 0) {
-                double value = featureValues[index];
-                for (Transformer t : e.getValue()) {
+        if(transformerMap.size() < featureNames.length) {
+            //
+            // We have transformers than feature names, so let's
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/impl/ArrayExample.java:420]: "have transformers" -> "have fewer transformers"

File: Core/src/main/java/org/tribuo/impl/ArrayExample.java [hunk 3]
```diff
@@ -408,14 +415,33 @@ public void set(Feature feature) {
...
+                    featureValues[index] = value;
+                }
+            }
+        } else {
+            //
+            // We have more transformers, so let's fetch them by name.
+            for(int i = 0; i < size; i++) {
+                List<Transformer> l = transformerMap.get(featureNames[i]);
+                if(l == null) {
+                    continue;
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/impl/ArrayExample.java:438]: I think I'd prefer this test to be `if (l != null)` and that way we can get rid of the continue statement.

File: Core/src/main/java/org/tribuo/transform/TransformTrainer.java [hunk 4]
```diff
@@ -81,10 +86,24 @@ public TransformTrainer(Trainer<T> innerTrainer, TransformationMap transformatio
 
     @Override
     public TransformedModel<T> train(Dataset<T> examples, Map<String, Provenance> instanceProvenance) {
+        
+        logger.info(String.format("%s %s", TransformTrainer.class.getName(), logger.getLevel()));
+        if(logger.isLoggable(Level.FINE)) {
+            logger.info(String.format("It says fine is loggable?"));
+        }
+        logger.fine(String.format("Creating transformers"));
+        logger.fine("Really doing it, though. Like for real");
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformTrainer.java:95]: I don't think this logging statement needs to be here.

File: Core/src/main/java/org/tribuo/transform/TransformTrainer.java [hunk 5]
```diff
@@ -81,10 +86,24 @@ public TransformTrainer(Trainer<T> innerTrainer, TransformationMap transformatio
...
+        for(Handler h : logger.getHandlers()) {
+            h.flush();
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformTrainer.java:97]: Does this flush need to be here either?

File: Core/src/main/java/org/tribuo/transform/TransformTrainer.java [hunk 6]
```diff
@@ -81,10 +86,24 @@ public TransformTrainer(Trainer<T> innerTrainer, TransformationMap transformatio
 
     @Override
     public TransformedModel<T> train(Dataset<T> examples, Map<String, Provenance> instanceProvenance) {
+        
+        logger.info(String.format("%s %s", TransformTrainer.class.getName(), logger.getLevel()));
+        if(logger.isLoggable(Level.FINE)) {
+            logger.info(String.format("It says fine is loggable?"));
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformTrainer.java:92]: Don't think this check needs to be here.

File: Core/src/main/java/org/tribuo/transform/TransformTrainer.java [hunk 7]
```diff
@@ -29,6 +29,9 @@
 
 import java.time.OffsetDateTime;
 import java.util.Map;
+import java.util.logging.Handler;
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformTrainer.java:32]: Not sure we need the Handler import once the debugging statements have been removed.

File: Core/src/main/java/org/tribuo/transform/TransformationMap.java [hunk 8]
```diff
@@ -49,9 +52,9 @@
 public class TransformationMap implements Configurable, Provenancable<ConfiguredObjectProvenance> {
 
     @Config(mandatory = true,description="Global transformations to apply after the feature specific transforms.")
-    private List<Transformation> globalTransformations;
+    private List<Transformation> globalTransformations = new ArrayList<>();
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformationMap.java:55]: This doesn't need to be assigned to an ArrayList if it's mandatory. It'll be overwritten by the config system, and it's overwritten in the constructor (though the constructor could do a defensive copy I guess).

File: Core/src/test/java/org/tribuo/transform/SimpleTransformTest.java [hunk 9]
```diff
@@ -232,7 +232,7 @@ public void testFeatureSpecific() {
         map.put("F6",Arrays.asList(SimpleTransform.exp(),SimpleTransform.log(),SimpleTransform.add(5)));
         map.put("F8",Arrays.asList(SimpleTransform.exp(),SimpleTransform.log()));
 
-        TransformationMap t = new TransformationMap(new ArrayList<>(),map);
+        TransformationMap t = new TransformationMap(new ArrayList<Transformation>(),map);
```
Comment (Craigacp) [inline Core/src/test/java/org/tribuo/transform/SimpleTransformTest.java:235]: Not sure this change needs to be here.

File: Core/src/main/java/org/tribuo/transform/TransformerMap.java [hunk 10]
```diff
@@ -129,16 +133,29 @@ public static double applyTransformerList(double value, List<Transformer> transf
...
         }
 
+        logger.fine(String.format("Transforming data set copy"));
+        
         newDataset.transform(this);
 
         return newDataset;
     }
+    
+    public int size() {
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformerMap.java:152]: Javadoc for the new methods please.

File: Core/src/main/java/org/tribuo/transform/TransformerMap.java [hunk 11]
```diff
@@ -129,16 +133,29 @@ public static double applyTransformerList(double value, List<Transformer> transf
...
+        return map.size();
+    }
+    
+    public List<Transformer> get(String featureName) {
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/transform/TransformerMap.java:156]: Javadoc.

**Review by Craigacp — CHANGES_REQUESTED** [review]
A few small things to fix.

File: Core/src/main/java/org/tribuo/impl/ArrayExample.java (hunk 1 above)
Comment (eelstretching) [reply Core/src/main/java/org/tribuo/impl/ArrayExample.java:164]: I presume you mean otherArr.size here.

File: Core/src/main/java/org/tribuo/transform/TransformTrainer.java (hunk 4 above)
Comment (eelstretching) [reply Core/src/main/java/org/tribuo/transform/TransformTrainer.java:95]: There were actually a few that didn't need to be there, I took them all out, including the flush in the next one.

**Comment by eelstretching** [comment]
Somehow I also managed to not add IDFTransformer to the original commit.

File: Core/src/main/java/org/tribuo/impl/ArrayExample.java (hunk 1 above)
Comment (Craigacp) [reply Core/src/main/java/org/tribuo/impl/ArrayExample.java]: Yeah, that one.

**Review by Craigacp — APPROVED** [review]
LGTM

**Pull request merged**